File: examples/video_to_docs.go
// Video-to-Docs: Full pipeline example (Go)
//
// Submits a video URL, polls for analysis, triggers AI rewrite,
// and downloads the result as markdown, DOCX, and PDF.
//
// Usage:
//
//	export DOCSIE_API_KEY="your_key"
//	go run video_to_docs.go https://example.com/video.mp4
//	go run video_to_docs.go https://example.com/video.mp4 sop
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

var (
	apiKey  = os.Getenv("DOCSIE_API_KEY")
	baseURL = envOrDefault("DOCSIE_BASE_URL", "https://app.docsie.io")
	api     = baseURL + "/api_v2/003"
	client  = &http.Client{Timeout: 60 * time.Second}
)

func envOrDefault(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

// ── HTTP helpers ─────────────────────────────────────────

func apiRequest(method, path string, body interface{}) (map[string]interface{}, error) {
	var reqBody io.Reader
	if body != nil {
		b, _ := json.Marshal(body)
		reqBody = bytes.NewReader(b)
	}
	req, err := http.NewRequest(method, api+path, reqBody)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Api-Key "+apiKey)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)
	return result, nil
}

func apiGet(path string) (map[string]interface{}, error) {
	return apiRequest("GET", path, nil)
}

func apiPost(path string, body map[string]interface{}) (map[string]interface{}, error) {
	return apiRequest("POST", path, body)
}

func poll(path string, timeoutSec, intervalSec int) (map[string]interface{}, error) {
	deadline := time.Now().Add(time.Duration(timeoutSec) * time.Second)
	for time.Now().Before(deadline) {
		data, err := apiGet(path)
		if err != nil {
			return nil, err
		}
		status := str(data, "status")
		if status == "" {
			status = str(data, "job_status")
		}
		switch status {
		case "done", "failed", "canceled":
			return data, nil
		}
		fmt.Printf("  ... %s\n", status)
		time.Sleep(time.Duration(intervalSec) * time.Second)
	}
	return nil, fmt.Errorf("timed out after %ds", timeoutSec)
}

func downloadFile(url, filename string) error {
	resp, err := http.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	f, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer f.Close()

	n, _ := io.Copy(f, resp.Body)
	fmt.Printf("  Downloaded: %s (%d KB)\n", filename, n/1024)
	return nil
}

func str(m map[string]interface{}, key string) string {
	if v, ok := m[key]; ok && v != nil {
		return fmt.Sprintf("%v", v)
	}
	return ""
}

func objMap(m map[string]interface{}, key string) map[string]interface{} {
	if v, ok := m[key].(map[string]interface{}); ok {
		return v
	}
	return map[string]interface{}{}
}

func arrLen(m map[string]interface{}, key string) int {
	if v, ok := m[key].([]interface{}); ok {
		return len(v)
	}
	return 0
}

// ── Main ─────────────────────────────────────────────────

func main() {
	if apiKey == "" {
		fmt.Fprintln(os.Stderr, "Error: Set DOCSIE_API_KEY environment variable")
		os.Exit(1)
	}
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "Usage: go run video_to_docs.go <video_url> [doc_style]")
		os.Exit(1)
	}

	videoURL := os.Args[1]
	docStyle := "guide"
	if len(os.Args) > 2 {
		docStyle = os.Args[2]
	}

	// 1. Submit
	fmt.Printf("==> Submitting video job (style=%s)...\n", docStyle)
	submitResp, err := apiPost("/video-to-docs/submit/", map[string]interface{}{
		"video_url":     videoURL,
		"quality":       "draft",
		"language":      "english",
		"doc_style":     docStyle,
		"auto_generate": false,
	})
	check(err)
	jobID := str(submitResp, "job_id")
	fmt.Printf("    Job ID: %s\n", jobID)

	// 2. Poll analysis
	fmt.Println("==> Polling analysis...")
	_, err = poll("/video-to-docs/"+jobID+"/status/", 900, 15)
	check(err)

	// 3. Get result
	fmt.Println("==> Fetching result...")
	result, err := apiGet("/video-to-docs/" + jobID + "/result/")
	check(err)
	markdown := str(result, "markdown")
	transcription := str(result, "transcription")
	fmt.Printf("    Markdown: %d chars\n", len(markdown))
	fmt.Printf("    Transcription: %d chars\n", len(transcription))
	fmt.Printf("    Sections: %d\n", arrLen(result, "sections"))
	fmt.Printf("    Images: %d\n", arrLen(result, "images"))

	os.WriteFile("analysis_result.md", []byte(markdown), 0644)
	os.WriteFile("transcription.txt", []byte(transcription), 0644)
	fmt.Println("    Saved: analysis_result.md, transcription.txt")

	// 4. Generate
	fmt.Println("==> Triggering AI rewrite...")
	genResp, err := apiPost("/video-to-docs/"+jobID+"/generate/", map[string]interface{}{
		"doc_style":      docStyle,
		"output_formats": []string{"md", "docx", "pdf"},
	})
	check(err)
	genJobID := str(genResp, "generate_job_id")
	fmt.Printf("    Generate job: %s\n", genJobID)

	fmt.Println("==> Polling generate job...")
	genResult, err := poll("/jobs/"+genJobID+"/", 300, 10)
	check(err)
	gen := objMap(genResult, "result")
	fmt.Printf("    Title: %s\n", str(gen, "title"))
	fmt.Printf("    Words: %s -> %s\n", str(gen, "input_word_count"), str(gen, "output_word_count"))

	os.WriteFile("generated.md", []byte(str(gen, "markdown")), 0644)
	fmt.Println("    Saved: generated.md")

	// 5. Download exports
	exports := objMap(gen, "exports")
	for _, fmt_ := range []string{"docx", "pdf"} {
		info := objMap(exports, fmt_)
		exportJobID := str(info, "job_id")
		if exportJobID == "" {
			continue
		}
		fmt.Printf("==> Waiting for %s export...\n", fmt_)
		exportResult, err := poll("/jobs/"+exportJobID+"/", 120, 5)
		check(err)
		exportData := objMap(exportResult, "result")
		url := str(exportData, "url")
		filename := str(exportData, "filename")
		if filename == "" {
			filename = "output." + fmt_
		}
		if url != "" {
			downloadFile(url, filename)
		}
	}

	fmt.Println("\n==> Complete!")
}

func check(err error) {
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
}

File: examples/video_to_docs.java
/**
 * Video-to-Docs: Full pipeline example (Java 11+)
 *
 * Submits a video URL, polls for analysis, triggers AI rewrite,
 * and downloads the result as markdown, DOCX, and PDF.
 *
 * Dependencies: com.google.code.gson:gson:2.11.0
 *
 * Compile & run:
 *   # Download gson jar (or use Maven/Gradle)
 *   curl -sL -o gson.jar https://repo1.maven.org/maven2/com/google/code/gson/gson/2.11.0/gson-2.11.0.jar
 *   export DOCSIE_API_KEY="your_key"
 *   javac -cp gson.jar video_to_docs.java
 *   java -cp .:gson.jar video_to_docs https://example.com/video.mp4
 *   java -cp .:gson.jar video_to_docs https://example.com/video.mp4 sop
 *
 * Or with Maven:
 *   <dependency>
 *     <groupId>com.google.code.gson</groupId>
 *     <artifactId>gson</artifactId>
 *     <version>2.11.0</version>
 *   </dependency>
 */

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;

import java.io.*;
import java.net.URI;
import java.net.http.*;
import java.nio.file.*;
import java.time.Duration;
import java.util.*;

public class video_to_docs {

    static final String API_KEY = System.getenv("DOCSIE_API_KEY");
    static final String BASE_URL = Optional.ofNullable(System.getenv("DOCSIE_BASE_URL"))
            .orElse("https://app.docsie.io");
    static final String API = BASE_URL + "/api_v2/003";
    static final HttpClient client = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(30))
            .build();
    static final Gson gson = new Gson();

    public static void main(String[] args) throws Exception {
        if (API_KEY == null || API_KEY.isEmpty()) {
            System.err.println("Error: Set DOCSIE_API_KEY environment variable");
            System.exit(1);
        }
        if (args.length < 1) {
            System.err.println("Usage: java video_to_docs <video_url> [doc_style]");
            System.exit(1);
        }

        String videoUrl = args[0];
        String docStyle = args.length > 1 ? args[1] : "guide";

        // 1. Submit
        System.out.printf("==> Submitting video job (style=%s)...%n", docStyle);
        JsonObject submitBody = new JsonObject();
        submitBody.addProperty("video_url", videoUrl);
        submitBody.addProperty("quality", "draft");
        submitBody.addProperty("language", "english");
        submitBody.addProperty("doc_style", docStyle);
        submitBody.addProperty("auto_generate", false);

        JsonObject submitResp = apiPost("/video-to-docs/submit/", submitBody);
        String jobId = submitResp.get("job_id").getAsString();
        System.out.printf("    Job ID: %s%n", jobId);

        // 2. Poll analysis
        System.out.println("==> Polling analysis...");
        poll("/video-to-docs/" + jobId + "/status/", 900, 15);

        // 3. Get result
        System.out.println("==> Fetching result...");
        JsonObject result = apiGet("/video-to-docs/" + jobId + "/result/");
        String markdown = getStr(result, "markdown");
        String transcription = getStr(result, "transcription");
        int sectionCount = result.has("sections") ? result.getAsJsonArray("sections").size() : 0;
        int imageCount = result.has("images") ? result.getAsJsonArray("images").size() : 0;

        System.out.printf("    Markdown: %d chars%n", markdown.length());
        System.out.printf("    Transcription: %d chars%n", transcription.length());
        System.out.printf("    Sections: %d%n", sectionCount);
        System.out.printf("    Images: %d%n", imageCount);

        Files.writeString(Path.of("analysis_result.md"), markdown);
        Files.writeString(Path.of("transcription.txt"), transcription);
        System.out.println("    Saved: analysis_result.md, transcription.txt");

        // 4. Generate
        System.out.println("==> Triggering AI rewrite...");
        JsonObject genBody = new JsonObject();
        genBody.addProperty("doc_style", docStyle);
        JsonArray formats = new JsonArray();
        formats.add("md"); formats.add("docx"); formats.add("pdf");
        genBody.add("output_formats", formats);

        JsonObject genResp = apiPost("/video-to-docs/" + jobId + "/generate/", genBody);
        String genJobId = genResp.get("generate_job_id").getAsString();
        System.out.printf("    Generate job: %s%n", genJobId);

        System.out.println("==> Polling generate job...");
        JsonObject genResult = poll("/jobs/" + genJobId + "/", 300, 10);
        JsonObject gen = genResult.has("result") ? genResult.getAsJsonObject("result") : new JsonObject();

        System.out.printf("    Title: %s%n", getStr(gen, "title"));
        System.out.printf("    Words: %s -> %s%n",
                gen.has("input_word_count") ? gen.get("input_word_count").getAsString() : "?",
                gen.has("output_word_count") ? gen.get("output_word_count").getAsString() : "?");

        Files.writeString(Path.of("generated.md"), getStr(gen, "markdown"));
        System.out.println("    Saved: generated.md");

        // 5. Download exports
        JsonObject exports = gen.has("exports") ? gen.getAsJsonObject("exports") : new JsonObject();
        for (String fmt : List.of("docx", "pdf")) {
            if (!exports.has(fmt)) continue;
            JsonObject info = exports.getAsJsonObject(fmt);
            String exportJobId = getStr(info, "job_id");
            if (exportJobId.isEmpty()) continue;

            System.out.printf("==> Waiting for %s export...%n", fmt.toUpperCase());
            JsonObject exportResult = poll("/jobs/" + exportJobId + "/", 120, 5);
            JsonObject exportData = exportResult.has("result")
                    ? exportResult.getAsJsonObject("result") : new JsonObject();
            String url = getStr(exportData, "url");
            String filename = getStr(exportData, "filename");
            if (filename.isEmpty()) filename = "output." + fmt;
            if (!url.isEmpty()) {
                downloadFile(url, filename);
            }
        }

        System.out.println("\n==> Complete!");
    }

    // ── HTTP helpers ─────────────────────────────────────

    static JsonObject apiGet(String path) throws Exception {
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(API + path))
                .header("Authorization", "Api-Key " + API_KEY)
                .header("Accept", "application/json")
                .timeout(Duration.ofSeconds(60))
                .GET().build();
        String body = client.send(req, HttpResponse.BodyHandlers.ofString()).body();
        return gson.fromJson(body, JsonObject.class);
    }

    static JsonObject apiPost(String path, JsonObject body) throws Exception {
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(API + path))
                .header("Authorization", "Api-Key " + API_KEY)
                .header("Content-Type", "application/json")
                .header("Accept", "application/json")
                .timeout(Duration.ofSeconds(60))
                .POST(HttpRequest.BodyPublishers.ofString(gson.toJson(body))).build();
        String resp = client.send(req, HttpResponse.BodyHandlers.ofString()).body();
        return gson.fromJson(resp, JsonObject.class);
    }

    static JsonObject poll(String path, int timeoutSec, int intervalSec) throws Exception {
        long deadline = System.currentTimeMillis() + timeoutSec * 1000L;
        while (System.currentTimeMillis() < deadline) {
            JsonObject data = apiGet(path);
            String status = getStr(data, "status");
            if (status.isEmpty()) status = getStr(data, "job_status");
            if (List.of("done", "failed", "canceled").contains(status)) return data;
            System.out.printf("  ... %s%n", status);
            Thread.sleep(intervalSec * 1000L);
        }
        throw new RuntimeException("Timed out after " + timeoutSec + "s");
    }

    static void downloadFile(String url, String filename) throws Exception {
        HttpRequest req = HttpRequest.newBuilder().uri(URI.create(url))
                .timeout(Duration.ofSeconds(60)).GET().build();
        byte[] bytes = client.send(req, HttpResponse.BodyHandlers.ofByteArray()).body();
        Files.write(Path.of(filename), bytes);
        System.out.printf("  Downloaded: %s (%d KB)%n", filename, bytes.length / 1024);
    }

    static String getStr(JsonObject obj, String key) {
        JsonElement el = obj.get(key);
        if (el == null || el.isJsonNull()) return "";
        return el.isJsonPrimitive() ? el.getAsString() : el.toString();
    }
}
